### PR TITLE
Delete unnecessary data for Funding Information task 

### PR DESCRIPTION
**Pull request by Harriethw**
# Context

we have several tasks which have branching logic within them.

E.g. on the funding-source page:

IF I answer personal savings
THEN I am taken to national insurance page
AND I do not see the identification page
or
IF I answer benefits
THEN I am taken to the identification page

If a user returns to the `funding-source` page and changes their answer
from `benefits` to `personalSavings`, then we need to ensure any answers
based on `benefits` are deleted.

Because of the way the page system works, it's tricky to get each page
to automate or handle this logic themselves. We only know what data
needs to be deleted /after/ the page data has been submitted.

I'm hesitant to introduce yet another controller function as we did with
the custom forms[1], because of the boilerplate overhead and the potential
for more complexity. This commit suggests that this data manipulation is
handled by the `save` function in the application service.

This `save` function is called by every standard page, and contains
other logic about how we handle the data. If this change is accepted, we
could go on to add more pages, such as equality and diversity, and
refactor the function to make it easier to add more pages.

[1] doc/architecture/decisions/0002_custom_forms.md

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Are any changes required to the e2e tests?
- [ ] If you've added a new route, have you added a new
  `auditEvent`? (see `server/routes/apply.ts` for examples)
- [ ] Are there environment variables or other infrastructure configuration which needs to be included in this release?
- [ ] Are there any data migrations required. Automatic or manual?
- [ ] Does this rely on changes being deployed to the CAS API?

## Post merge checklist

Once we've merged it will be auto-deployed to the dev environment.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-community-accommodation-tier-2-ui).

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

**Comment by Harriethw** [comment]
https://trello.com/c/hN7fVtbR/594-delete-equality-and-diversity-data-if-application-chooses-not-to-answer-the-questions

File: server/form-pages/apply/area-and-funding/funding-information/fundingInformation.ts [hunk 1]
```diff
@@ -51,6 +51,8 @@ export default class FundingSource implements TaskListPage {
 
   next() {
     if (this.body.fundingSource === 'personalSavings') {
+      // it needs to update itself
+      // and delete/set to null keys for other pages, identification, alternative-identification
```
Comment (patrickjfl) [inline server/form-pages/apply/area-and-funding/funding-information/fundingInformation.ts:55]: I assume these comments are hangovers from a previous implementation idea?

**Review by patrickjfl — APPROVED** [review]
I'm happy with this as an approach, and I agree this should be the responsibility of the `save` function. Maybe as we extend this to other cases we'll see opportunities to refactor. Nice work 🙌

File: server/services/applicationService.ts [hunk 2]
```diff
@@ -103,6 +103,15 @@ export default class ApplicationService {
     return applicationData
   }
 
+  private removeConditionalData(applicationData: AnyValue): AnyValue {
```
Comment (edavey) [inline server/services/applicationService.ts:106]: these changes definitely help! 

Would this be clearer if the function name reflected the fact that we're only clearing "conditional" answers where their follow-on question is no longer applicable? e.g.

`removeStaleFollowOnAnswers()` or
`clearOrphanedFollowOnAnswers()`

**Review by edavey — APPROVED** [review]
👍

File: server/services/applicationService.ts [hunk 3]
```diff
@@ -64,26 +60,47 @@ export default class ApplicationService {
...
 
-      const checkYourAnswersTaskName = getTaskName(CheckYourAnswers)
-      const checkYourAnswersPageName = getPageName(CheckYourAnswers)
+  private removeCheckYourAnswersIfPageChange(
+    applicationData: AnyValue,
+    pageName: string,
+    oldBody: AnyValue,
+    newBody: AnyValue,
+  ) {
+    const checkYourAnswersTaskName = getTaskName(CheckYourAnswers)
```
[the remark this answers is not in the document]
Comment (Harriethw) [reply server/services/applicationService.ts:90]: yeahh, I agree it's not the clearest, but not sure what alternatives would be better. `taskNameForCheckYourAnswers` ? 😬

File: server/form-pages/apply/area-and-funding/funding-information/fundingInformation.ts (hunk 1 above)
Comment (Harriethw) [reply server/form-pages/apply/area-and-funding/funding-information/fundingInformation.ts:55]: yep good spot!

File: server/services/applicationService.ts [hunk 4]
```diff
@@ -64,26 +60,47 @@ export default class ApplicationService {
...
+  ): AnyValue {
+    const newApplicationData = applicationData || {}
+    newApplicationData[taskName] = newApplicationData[taskName] || {}
+    newApplicationData[taskName][pageName] = page.body
+    return newApplicationData
+  }
 
-      const checkYourAnswersTaskName = getTaskName(CheckYourAnswers)
-      const checkYourAnswersPageName = getPageName(CheckYourAnswers)
+  private removeCheckYourAnswersIfPageChange(
```
[the remark this answers is not in the document]
Comment (Harriethw) [reply server/services/applicationService.ts:84]: hmm, I agree it's not as nice as I'd like. But wouldn't wrapping it in `setCheckYourAnswersAsIncomplete` obscure things more, as we lose the clarification that we're doing this only if there's a page change? And then referring to `Incomplete` means we'd be bringing in terms from elsewhere in the UI that the `ApplicationService` doesn't necessarily know about. 
Would using the term `delete` be more descriptive, as at least that describes the literal transformation we're applying to the data?
e.g. 
`deleteOrphanedFollowOnAnswers`
`deleteCheckYourAnswersIfPageChange`

File: server/services/applicationService.ts (hunk 2 above)
Comment (Harriethw) [reply server/services/applicationService.ts:106]: good point... how do you feel about `deleteOrphanedFollowOnAnswers`? https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/pull/284#discussion_r1384706514

File: server/services/applicationService.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Harriethw) [reply server/services/applicationService.ts:84]: @patrickjfl I'm going to merge after lunch to move this along, but happy to keep chatting about potential refactors as I will be revisiting to implement further tasks that have conditional data.

**Pull request merged**